File: libs/moqtail-ts/src/model/data/fetch_object.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { KeyValuePair } from '../common/pair'
import { ObjectStatus } from './constant'
import { Location } from '../common/location'

export class FetchObject {
  public readonly location: Location
  public readonly subgroupId: bigint

  private constructor(
    location: Location,
    subgroupId: bigint | number,
    public readonly publisherPriority: number,
    public readonly extensionHeaders: KeyValuePair[] | null,
    public readonly objectStatus: ObjectStatus | null,
    public readonly payload: Uint8Array | null,
  ) {
    this.location = location
    this.subgroupId = BigInt(subgroupId)
  }

  get groupId(): bigint {
    return this.location.group
  }
  get objectId(): bigint {
    return this.location.object
  }

  static newWithStatus(
    groupId: bigint | number,
    subgroupId: bigint | number,
    objectId: bigint | number,
    publisherPriority: number,
    extensionHeaders: KeyValuePair[] | null,
    objectStatus: ObjectStatus,
  ): FetchObject {
    return new FetchObject(
      new Location(groupId, objectId),
      subgroupId,
      publisherPriority,
      extensionHeaders,
      objectStatus,
      null,
    )
  }

  static newWithPayload(
    groupId: bigint | number,
    subgroupId: bigint | number,
    objectId: bigint | number,
    publisherPriority: number,
    extensionHeaders: KeyValuePair[] | null,
    payload: Uint8Array,
  ): FetchObject {
    return new FetchObject(
      new Location(groupId, objectId),
      subgroupId,
      publisherPriority,
      extensionHeaders,
      null,
      payload,
    )
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(this.location.group)
    buf.putVI(this.subgroupId)
    buf.putVI(this.location.object)
    buf.putU8(this.publisherPriority)
    const extensionHeaders = new ByteBuffer()
    if (this.extensionHeaders) {
      for (const header of this.extensionHeaders) {
        extensionHeaders.putKeyValuePair(header)
      }
    }
    const extBytes = extensionHeaders.toUint8Array()
    buf.putLengthPrefixedBytes(extBytes)
    if (this.payload) {
      buf.putLengthPrefixedBytes(this.payload)
    } else {
      buf.putVI(0)
      buf.putVI(this.objectStatus!)
    }
    return buf.freeze()
  }

  static deserialize(buf: BaseByteBuffer): FetchObject {
    const groupId = buf.getVI()
    const subgroupId = buf.getVI()
    const objectId = buf.getVI()
    const publisherPriority = buf.getU8()
    const extLen = buf.getNumberVI()
    let extensionHeaders: KeyValuePair[] | null = null
    if (extLen > 0) {
      const headerBytes = new FrozenByteBuffer(buf.getBytes(extLen))
      extensionHeaders = []
      while (headerBytes.remaining > 0) {
        extensionHeaders.push(headerBytes.getKeyValuePair())
      }
    }
    const payloadLen = buf.getNumberVI()
    let objectStatus: ObjectStatus | null = null
    let payload: Uint8Array | null = null
    if (payloadLen === 0) {
      objectStatus = ObjectStatus.tryFrom(buf.getVI())
    } else {
      payload = buf.getBytes(payloadLen)
    }
    return new FetchObject(
      new Location(groupId, objectId),
      subgroupId,
      publisherPriority,
      extensionHeaders,
      objectStatus,
      payload,
    )
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('FetchObject', () => {
    test('roundtrip', () => {
      const groupId = 9n
      const subgroupId = 144n
      const objectId = 10n
      const publisherPriority = 255
      const extensionHeaders = [
        KeyValuePair.tryNewVarInt(0, 1000),
        KeyValuePair.tryNewBytes(9, new TextEncoder().encode('wololoo')),
      ]
      const payload = new TextEncoder().encode('01239gjawkk92837aldmi')
      const fetchObject = FetchObject.newWithPayload(
        groupId,
        subgroupId,
        objectId,
        publisherPriority,
        extensionHeaders,
        payload,
      )
      const frozen = fetchObject.serialize()
      const parsed = FetchObject.deserialize(frozen)
      expect(parsed.groupId).toBe(groupId)
      expect(parsed.subgroupId).toBe(subgroupId)
      expect(parsed.objectId).toBe(objectId)
      expect(parsed.publisherPriority).toBe(publisherPriority)
      expect(parsed.extensionHeaders).toEqual(extensionHeaders)
      expect(parsed.payload).toEqual(payload)
      expect(frozen.remaining).toBe(0)
    })
    test('excess roundtrip', () => {
      const groupId = 9n
      const subgroupId = 144n
      const objectId = 10n
      const publisherPriority = 255
      const extensionHeaders = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('wololoo')),
      ]
      const payload = new TextEncoder().encode('01239gjawkk92837aldmi')
      const fetchObject = FetchObject.newWithPayload(
        groupId,
        subgroupId,
        objectId,
        publisherPriority,
        extensionHeaders,
        payload,
      )
      const serialized = fetchObject.serialize().toUint8Array()
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      const excess = new Uint8Array([9, 1, 1])
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const parsed = FetchObject.deserialize(frozen)
      expect(parsed.groupId).toBe(groupId)
      expect(parsed.subgroupId).toBe(subgroupId)
      expect(parsed.objectId).toBe(objectId)
      expect(parsed.publisherPriority).toBe(publisherPriority)
      expect(parsed.extensionHeaders).toEqual(extensionHeaders)
      expect(parsed.payload).toEqual(payload)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })
    test('partial message fails', () => {
      const groupId = 9n
      const subgroupId = 144n
      const objectId = 10n
      const publisherPriority = 255
      const extensionHeaders = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('wololoo')),
      ]
      const payload = new TextEncoder().encode('01239gjawkk92837aldmi')
      const fetchObject = FetchObject.newWithPayload(
        groupId,
        subgroupId,
        objectId,
        publisherPriority,
        extensionHeaders,
        payload,
      )
      const serialized = fetchObject.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        FetchObject.deserialize(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/data/fetch_object.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::model::common::pair::KeyValuePair;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use bytes::{Buf, BufMut, Bytes, BytesMut};

use super::constant::ObjectStatus;

#[derive(Debug, Clone, PartialEq)]
pub struct FetchObject {
  pub group_id: u64,
  pub subgroup_id: u64,
  pub object_id: u64,
  pub publisher_priority: u8,
  pub extension_headers: Option<Vec<KeyValuePair>>,
  pub object_status: Option<ObjectStatus>,
  pub payload: Option<Bytes>,
}

impl FetchObject {
  pub fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();

    buf.put_vi(self.group_id)?;
    buf.put_vi(self.subgroup_id)?;
    buf.put_vi(self.object_id)?;
    buf.put_u8(self.publisher_priority);

    if let Some(ext_headers) = &self.extension_headers {
      let mut ext_buf = BytesMut::new();
      for header in ext_headers {
        ext_buf.extend_from_slice(&header.serialize()?);
      }
      buf.put_vi(ext_buf.len() as u64)?;
      buf.extend_from_slice(&ext_buf);
    } else {
      buf.put_vi(0u64)?;
    }

    if let Some(status) = self.object_status {
      buf.put_vi(0u64)?;
      buf.put_vi(status)?;
    } else if let Some(payload) = &self.payload {
      buf.put_vi(payload.len() as u64)?;
      buf.extend_from_slice(payload);
    } else {
      return Err(ParseError::ProtocolViolation {
        context: "FetchObject::serialize",
        details: "No object status, no payload".to_string(),
      });
    }

    Ok(buf.freeze())
  }

  pub fn deserialize(bytes: &mut Bytes) -> Result<Self, ParseError> {
    let group_id = bytes.get_vi()?;
    let subgroup_id = bytes.get_vi()?;
    let object_id = bytes.get_vi()?;

    if bytes.remaining() < 1 {
      return Err(ParseError::NotEnoughBytes {
        context: "FetchObject::deserialize",
        needed: 1,
        available: 0,
      });
    }
    let publisher_priority = bytes.get_u8();

    let ext_len = bytes.get_vi()?;
    let ext_len: usize =
      ext_len
        .try_into()
        .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
          context: "ObjectDatagram::deserialize",
          from_type: "u64",
          to_type: "usize",
          details: e.to_string(),
        })?;
    let extension_headers = if ext_len > 0 {
      if bytes.remaining() < ext_len {
        return Err(ParseError::NotEnoughBytes {
          context: "ObjectDatagram::deserialize",
          needed: ext_len,
          available: bytes.remaining(),
        });
      }
      let mut header_bytes = bytes.copy_to_bytes(ext_len);
      let mut headers: Vec<KeyValuePair> = Vec::new();
      while header_bytes.has_remaining() {
        let h = KeyValuePair::deserialize(&mut header_bytes).map_err(|_| {
          ParseError::ProtocolViolation {
            context: "ObjectDatagram::deserialize(headers)",
            details: "Can't parse headers".to_string(),
          }
        })?;
        headers.push(h);
      }
      Some(headers)
    } else {
      None
    };

    let payload_len = bytes.get_vi()?;

    let (payload, object_status) = if payload_len == 0 {
      let status_raw = bytes.get_vi()?;
      let status = ObjectStatus::try_from(status_raw)?;
      (None, Some(status))
    } else {
      let payload_len: usize = payload_len
        .try_into()
        .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
          context: "ObjectDatagram::deserialize",
          from_type: "u64",
          to_type: "usize",
          details: e.to_string(),
        })?;
      if bytes.remaining() < payload_len {
        return Err(ParseError::NotEnoughBytes {
          context: "ObjectDatagram::deserialize",
          needed: ext_len,
          available: bytes.remaining(),
        });
      }
      (Some(bytes.copy_to_bytes(payload_len)), None)
    };

    Ok(FetchObject {
      group_id,
      subgroup_id,
      object_id,
      publisher_priority,
      extension_headers,
      object_status,
      payload,
    })
  }
}

#[cfg(test)]
mod tests {

  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let group_id: u64 = 9;
    let subgroup_id = 144;
    let object_id: u64 = 10;
    let publisher_priority: u8 = 255;
    let extension_headers = Some(vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"wololoo")).unwrap(),
    ]);
    let object_status = None;
    let payload = Some(Bytes::from_static(b"01239gjawkk92837aldmi"));

    let fetch_object = FetchObject {
      group_id,
      subgroup_id,
      object_id,
      publisher_priority,
      extension_headers,
      payload,
      object_status,
    };

    let mut buf = fetch_object.serialize().unwrap();
    let deserialized = FetchObject::deserialize(&mut buf).unwrap();
    assert_eq!(deserialized, fetch_object);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let group_id: u64 = 9;
    let subgroup_id = 144;
    let object_id: u64 = 10;
    let publisher_priority: u8 = 255;
    let extension_headers = Some(vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"wololoo")).unwrap(),
    ]);
    let object_status = None;
    let payload = Some(Bytes::from_static(b"01239gjawkk92837aldmi"));

    let fetch_object = FetchObject {
      group_id,
      subgroup_id,
      object_id,
      publisher_priority,
      extension_headers,
      payload,
      object_status,
    };

    let serialized = fetch_object.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let deserialized = FetchObject::deserialize(&mut buf).unwrap();
    assert_eq!(deserialized, fetch_object);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let group_id: u64 = 9;
    let subgroup_id = 144;
    let object_id: u64 = 10;
    let publisher_priority: u8 = 255;
    let extension_headers = Some(vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"wololoo")).unwrap(),
    ]);
    let object_status = None;
    let payload = Some(Bytes::from_static(b"01239gjawkk92837aldmi"));

    let fetch_object = FetchObject {
      group_id,
      subgroup_id,
      object_id,
      publisher_priority,
      extension_headers,
      payload,
      object_status,
    };
    let buf = fetch_object.serialize().unwrap();
    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = FetchObject::deserialize(&mut partial);
    assert!(deserialized.is_err());
  }
}
